core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -0,0 +1,713 @@
+    }
+
+    public enum State
+    {
+        /**
+         * Stage is planned but has not been scheduled yet.  A stage will
+         * be in the planned state until, the dependencies of the stage
+         * have begun producing output.
+         */
+        PLANNED(false, false),
+        /**
+         * Stage tasks are being scheduled on nodes.
+         */
+        SCHEDULING(false, false),
+        /**
+         * All stage tasks have been scheduled, but splits are still being scheduled.
+         */
+        SCHEDULING_SPLITS(false, false),
+        /**
+         * Stage has been scheduled on nodes and ready to execute, but all tasks are still queued.
+         */
+        SCHEDULED(false, false),
+        /**
+         * Stage is running.
+         */
+        RUNNING(false, false),
+        /**
+         * Stage has finished executing and output being consumed.
+         * In this state, at-least one of the tasks is flushing and the non-flushing tasks are finished
+         */
+        FLUSHING(false, false),
+        /**
+         * Stage has finished executing and all output has been consumed.
+         */
+        FINISHED(true, false),
+        /**
+         * Stage was canceled by a user.
+         */
+        CANCELED(true, false),
+        /**
+         * Stage was aborted due to a failure in the query.  The failure
+         * was not in this stage.
+         */
+        ABORTED(true, true),
+        /**
+         * Stage execution failed.
+         */
+        FAILED(true, true);
+
+        private final boolean doneState;
+        private final boolean failureState;
+
+        State(boolean doneState, boolean failureState)
+        {
+            checkArgument(!failureState || doneState, "%s is a non-done failure state", name());
+            this.doneState = doneState;
+            this.failureState = failureState;
+        }
+
+        /**
+         * Is this a terminal state.
+         */
+        public boolean isDone()
+        {
+            return doneState;
+        }
+
+        /**
+         * Is this a non-success terminal state.
+         */
+        public boolean isFailure()
+        {
+            return failureState;
+        }
+
+        public boolean canScheduleMoreTasks()
+        {
+            switch (this) {
+                case PLANNED:
+                case SCHEDULING:
+                    // workers are still being added to the query
+                    return true;
+                case SCHEDULING_SPLITS:
+                case SCHEDULED:
+                case RUNNING:
+                case FLUSHING:
+                case FINISHED:
+                case CANCELED:
+                    // no more workers will be added to the query
+                    return false;
+                case ABORTED:
+                case FAILED:
+                    // DO NOT complete a FAILED or ABORTED stage.  This will cause the
```

Comment:
I see this just copied - but I have no clue how this comment relates to this code location.

Reply:
To be hones I also don't really understand this comment. I just moved the entire method.